src/main/java/com/google/devtools/build/lib/exec/RunfilesTreeUpdater.java

```diff
@@ -0,0 +1,159 @@
+   * mapping will be removed.
+   */
+  private final ConcurrentMap<PathFragment, AtomicInteger> monitorsWithRefcnt
+      = new ConcurrentHashMap<>();
+  private final boolean enableRunfiles;
+
+
+  public RunfilesTreeUpdater(boolean enableRunfiles) {
+    this.enableRunfiles = enableRunfiles;
+  }
+
+  private static void updateRunfilesTree(Path execRoot, PathFragment runfilesDir,
+      BinTools binTools, ImmutableMap<String, String> env, OutErr outErr, boolean enableRunfiles)
+      throws IOException, ExecException {
+    Path runfilesDirPath = execRoot.getRelative(runfilesDir);
+    Path inputManifest = RunfilesSupport.inputManifestPath(runfilesDirPath);
+    if (!inputManifest.exists()) {
+      // TODO(buchgr): Add logic to create a runfiles tree if --nobuild_runfile_manifests is set
+      throw new UserExecException("Can't build local runfiles tree because "
+          + "--nobuild_runfile_manifests is set.");
+    }
+
+    Path outputManifest = RunfilesSupport.outputManifestPath(runfilesDirPath);
+    try {
+      // Avoid rebuilding the runfiles directory if the manifest in it matches the input manifest,
+      // implying the symlinks exist and are already up to date. If the output manifest is a
+      // symbolic link, it is likely a symbolic link to the input manifest, so we cannot trust it as
+      // an up-to-date check.
+      FileStatus stat = outputManifest.stat();
+      if (stat.isFile() && !stat.isSymbolicLink() && Arrays.equals(outputManifest.getDigest(),
+          inputManifest.getDigest())) {
+        return;
+      }
+    } catch (IOException e) {
+      // Ignore it - we will just try to create runfiles directory.
+    }
+
+    if (!runfilesDirPath.exists()) {
+      runfilesDirPath.createDirectoryAndParents();
+    }
+
+    SymlinkTreeHelper helper = new SymlinkTreeHelper(inputManifest, runfilesDirPath,
+        /* filesetTree= */ false);
+    helper.createSymlinks(execRoot, outErr, binTools, env, enableRunfiles);
+  }
+  
+  private AtomicInteger tryGetMonitor(PathFragment runfilesDirectory) {
+    AtomicInteger newMonitorWithRefcnt = new AtomicInteger(1);
+    AtomicInteger monitorWithRefcnt =
+        monitorsWithRefcnt.putIfAbsent(runfilesDirectory, newMonitorWithRefcnt);
+    if (monitorWithRefcnt == null) {
+      // This call created the entry for the runfiles directory as there wasn't one before.
+      return newMonitorWithRefcnt;
+    }
+    while (true) {
+      int count = monitorWithRefcnt.get();
+      if (count == 0) {
+        // Getting the lock failed as the reference count was 0. A value of 0 means that we
+        // raced with returnMonitor() and the monitor is about to be removed from the
+        // monitorsWithRefcnt map.
+        return null;
+      }
+      if (monitorWithRefcnt.compareAndSet(count, count + 1)) {
+        return monitorWithRefcnt;
+      }
+    }
+  }
+
+  private AtomicInteger getMonitor(PathFragment runfilesDirectory) {
```

Comment:
I think this works, but I'm kinda afraid of this kind of multi-threaded code. Do you have any other alternatives that do not require this? For example (off the top of my head, I haven't given this too much thought), you could synchronize on the `Path` object referring to one of the manifests (not an `Artifact`, though, that's not guaranteed to be unique)

Or you could just add the lock to the `TestRunnerAction`. That has the advantage of not abusing the monitor of a `Path`. Of course, this only works if this is the only case when a runfiles tree is reused (I don't know what happens when the same binary and its runfiles tree is used as a tool in two local genrules executing in parallel)

Reply:
> For example (off the top of my head, I haven't given this too much thought), you could synchronize on the Path object referring to one of the manifests (not an Artifact, though, that's not guaranteed to be unique)

Exactly. What this code does is map equal path objects to a unique object. I believe it's the most principled way of solving this problem, in particular because it pushes the complexity down the stack where it belongs.

> Or you could just add the lock to the TestRunnerAction. 

I think this would serialize execution of shards?

> I don't know what happens when the same binary and its runfiles tree is used as a tool in two local genrules executing in parallel

What would happen is exactly what this code is trying to prevent :).

Reply:
Let me word this more carefully: you are probably thinking `public synchronized runTest() { }`, I'm thinking

```
Object runfilesLock;

public void runTest() {
  if (runfilesTreeNeeded()) {
    synchronized (runfilesLock) {
      ...
    }
  }
}
```

...and what would that be in case of the two-genrules-using-the-same-tool situation?

Reply:
@lberki 

This would serialize runfiles tree creation for all tests where runfilesTreeNeeded() returns false. This is a regression.

>...and what would that be in case of the two-genrules-using-the-same-tool situation?

As my change description states runfiles tree creation when --nobuild_runfile_links is true does not work for genrules. It only works for tests (because there is custom code). A side effect of this change is that it will make it work for genrules as well.